Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -0,0 +1,418 @@
+                self.subscribedTopicsPointer,
+                topic,
+                KafkaPartition.unassigned.rawValue
+            )
+        }
+
+        let result = self.client.withKafkaHandlePointer { handle in
+            rd_kafka_subscribe(handle, subscribedTopicsPointer)
+        }
+
+        guard result.rawValue == 0 else {
+            throw KafkaError(rawValue: result.rawValue)
+        }
+    }
+
+    /// Assign the``KafkaConsumer`` to a specific `partition` of a `topic`.
+    /// - Parameter topic: Name of the topic that this ``KafkaConsumer`` will read from.
+    /// - Parameter partition: Partition that this ``KafkaConsumer`` will read from.
+    /// - Parameter offset: The topic offset where reading begins. Defaults to the offset of the last read message.
+    private func assign(
+        topic: String,
+        partition: KafkaPartition,
+        offset: Int64 = Int64(RD_KAFKA_OFFSET_END)
+    ) throws {
+        // TODO: is this state needed for a method that is only invoked upon init?
+        switch self.state {
+        case .started:
+            break
+        case .closed:
+            throw KafkaError(description: "Trying to invoke method on consumer that has been closed.")
+        }
+
+        guard let partitionPointer = rd_kafka_topic_partition_list_add(
+            self.subscribedTopicsPointer,
+            topic,
+            partition.rawValue
+        ) else {
+            fatalError("rd_kafka_topic_partition_list_add returned invalid pointer")
+        }
+
+        partitionPointer.pointee.offset = offset
+
+        let result = self.client.withKafkaHandlePointer { handle in
+            rd_kafka_assign(handle, self.subscribedTopicsPointer)
+        }
+
+        guard result == RD_KAFKA_RESP_ERR_NO_ERROR else {
+            throw KafkaError(rawValue: result.rawValue)
+        }
+    }
+
+    /// Receive new messages and forward the result to the ``messages`` `AsyncSequence`.
+    func produceMore() async {
+        let messageresult: Element
+        do {
+            guard let message = try await self.poll() else {
+                return
+            }
+            messageresult = .success(message)
+        } catch {
+            messageresult = .failure(error)
+        }
+
+        let yieldresult = self.messagesSource.yield(messageresult)
+        switch yieldresult {
+        case .produceMore:
+            await self.produceMore()
+        case .dropped, .stopProducing:
+            return
+        }
+    }
+
+    /// Request a new message from the Kafka cluster.
+    /// This method blocks for a maximum of `timeout` milliseconds.
+    /// - Parameter timeout: Maximum amount of milliseconds this method waits for a new message.
+    /// - Returns: A ``KafkaConsumerMessage`` or `nil` if there are no new messages.
+    func poll(timeout: Int32 = 100) async throws -> KafkaConsumerMessage? {
+        // TODO: clock API
+        // TODO: ideally: make private
+        // TODO: is this state needed here?
```

Comment:
1. This is a good candidate for `Clock API` — shall I create a ticket?
2. Again, this function is not meant to be used from outside the `KafkaConsumer` — we could therefore think about omitting `state` here as well